DOCKERFILE + karma changes for installation on Windows

EricZLou opened the pull request:
## Explanation
**Dockerfile allows for reliable installation of Oppia on Windows.**
**Installation process for non-Windows users will remain the same as before.** 
- install_prerequisites.sh works
- start.sh works
- run_frontend_tests works 50% of the time. Details in the Google Doc below.
- run_e2e_tests has not been tested yet, but this PR is mainly to get the other 3 components in the codebase. 

**Full documentation on using docker is included here (click on August):** 
https://docs.google.com/document/d/1TObGklQDj3tXD3jWA7kM2SjFlpVA6AMO57EbjbB7xlA/edit?usp=sharing

**Full instructions for operating Docker for Oppia:**
https://docs.google.com/document/d/1tMLuTNap6nHFerTbqfyYUQc8sXSUxzxWths7mK5D8-c/edit?usp=sharing

**Changes also made to karma.conf.ts and package.json**

**TODO**
- Try to figure out what's causing the race condition for Karma only starting half the time
- Rewrite Oppia installation on Windows Wiki

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

EricZLou committed:
Dockerfile, karma.conf.ts, package.json

EricZLou commented:
Not sure which Changelog or Project this should go under. Please let me know!

kevinlee12 commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+   apt-get install -y vim
+
+# Install Python and OpenJDK-8
+RUN apt-get update && \
```
Instead of doing this, can we run this file instead? https://github.com/oppia/oppia/blob/develop/scripts/install_prerequisites.sh

kevinlee12 replied:
That way, it's all a single source of truth

renana replied:
Not really, because at this stage the Oppia folder is not available yet and these are needed for setting other things like the Chrome installation and other tools that are not included in the script.

kevinlee12 replied:
Okay, in that case, please do reference that file since both files need to change in case we need to change the packages in either.

EricZLou replied:
I believe we can replace the first apt-get install line (apt-get install -y curl python-setuptools git python-dev python-pip python-yaml) with the prerequisites script. As long as we mkdir /home/oppia early on, and copy over the install_prerequisites script, we can get that going. The rest of the apt-get statements will need to stay there, however.

EricZLou replied:
Putting in an edit soon with a slightly reorganized Dockerfile that calls install_prerequisites.

kevinlee12 replied:
That is fine, my concern is preventing discrepancies in the prerequisite packages. At the very least, reference that `install_prerequisites.sh` script.

seanlip replied:
If we don't actually use install_prerequisites, then we'd need a backlink as well from install_prerequisites to Dockerfile, right?

So maybe just including it is better, if possible.

EricZLou replied:
Done. Added to Dockerfile.

kevinlee12 reviewed (changes requested):
Hi @ezl-13, thanks for the PR! I left a comment.

seanlip commented on Dockerfile:
```diff
@@ -0,0 +1,49 @@
+FROM ubuntu:latest
```
Please add the usual file headers (Apache license etc.), similar to other files in scripts/.

Also, if you look at most of the Oppia scripts, there is a long descriptive comment at the top explaining how the file should be used. I would suggest that we do that here as well. Explain any limitations here, too.

EricZLou replied:
Will do later tonight. Thanks Sean

EricZLou replied:
Will have to do this and the isDocker explanations tomorrow!

seanlip replied:
I think this is done -- in the future, please make sure to reply "Done" to a comment thread when it's done, and resolve it. Thanks!

EricZLou replied:
Yep. This is done. I'll make sure to do that!

seanlip commented on core/tests/karma.conf.ts:
```diff
@@ -91,7 +92,7 @@ module.exports = function(config) {
       }
     },
-    autoWatch: true,
+    autoWatch: isDocker ? false : true,
```
Add a comment above this line explaining the rationale.

EricZLou replied:
Reverted.

seanlip commented on core/tests/karma.conf.ts:
```diff
@@ -105,8 +106,12 @@ module.exports = function(config) {
       Chrome_Travis: {
-        base: 'Chrome',
-        flags: ['--no-sandbox']
+        base: isDocker ? 'ChromeHeadless' : 'Chrome',
```
Add comments above this and the next line explaining the rationale.

EricZLou replied:
Done.

seanlip reviewed (changes requested):
Thanks @ezl-13! I took a pass.

seanlip commented:
Also, heads-up ... tests seem to be failing.

(For codeowners, I think you can put this under @kevinlee12 and yourself.)

EricZLou committed:
Dockerfile calls install_prerequisites.sh

EricZLou commented:
Updated the Dockerfile to include install_prerequisites.sh but have not addressed the other comments yet.

EricZLou committed:
Merge remote-tracking branch 'upstream/develop' into docker

EricZLou committed:
Merge branch 'docker' of https://github.com/ezl-13/oppia into docker

EricZLou committed:
Add instructions to Dockerfile, revert autoWatch, describe isDocker

EricZLou commented:
- Added instructions and headers to Dockerfile
- Described browser flags in karma.conf.ts
- Added me and @kevinlee12 to Dockerfile CODEOWNERS

PTAL!

kevinlee12 commented:
@ezl-13, the e2e tests are failing, ptal!

seanlip commented on Dockerfile:
```diff
@@ -0,0 +1,161 @@
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
```
2019

EricZLou replied:
Oops. I just copied and pasted that. Good catch.

seanlip commented on core/tests/karma.conf.ts:
```diff
@@ -105,8 +106,15 @@ module.exports = function(config) {
-        flags: ['--no-sandbox']
+        // Karma can only connect to ChromeHeadless when inside Docker.
+        base: isDocker ? 'ChromeHeadless' : 'Chrome',
+        // --no-sandbox and --disable-setuid-sandbox perform similar functions
```
Could you explain what the flags actually do?

EricZLou replied:
To be completely honest, I really am only 30% knowledgeable about why these flags are needed (besides --no-sandbox). Online, on all the troubleshooting forums for Karma not starting with ChromeHeadless, almost all of the solutions state to use --disable-setuid-sandbox and --disable-web-security in conjunction with --no-sandbox. 

I'm not sure why this combination of flags allows Karma to work, and people online don't explain it. I've looked into what the flags individually do, though I can't tell why they are needed in a Dockerized environment. 

Will it be enough just to explain what the flags do, even if it still isn't clear why the particular flag is needed in this scenario?

seanlip replied:
In that case, I suggest providing a link to an authoritative site that explains this, so that a developer can look at it too for context.

EricZLou replied:
Got it. Done!

seanlip commented on Dockerfile:
```diff
@@ -0,0 +1,161 @@
+ENV PATH $PATH:$JAVA_HOME/bin
+RUN export JAVA_HOME
+
+# Install dumb-init (Signal handling of SIGINT/SIGTERM/SIGKILL etc.)
```
Can you elaborate a bit more on this comment? I.e. explain exactly what dumb-init is needed for and what would go wrong without it. It's OK to include details, don't just say "etc.".

This is useful context for future developers.

EricZLou replied:
Good point. Done.

seanlip reviewed (commented):
Thanks! Couple more notes.

EricZLou commented:
**Note:** 
Travis is failing because the Dockerfile is conflicting with GAE config. 
https://travis-ci.org/oppia/oppia/jobs/568761663#L2157
**Reason:**
Google App Engine requires that whenever there is a Dockerfile in the directory, the runtime must be set to custom. Currently, the runtime is set to python27. Our Dockerfile’s base image is ubuntu. So, if we set the runtime to custom, GAE never knows that it needs to use python27 since neither the GAE config nor our Dockerfile would specify python27.
The solution that I am using in the Docker container to avoid this problem when GAE runs is to `rm Dockerfile` when creating the container. One solution to Travis tests hitting this error would be to have a check like `if isDocker: run rm Dockerfile` before starting up GAE.

seanlip commented:
/cc @LanJosh

seanlip commented on Dockerfile:
```diff
@@ -0,0 +1,161 @@
+#   ex: docker build -t oppia_image . --no-cache
+#
+# STEP 4
+# Follow the instructions on how to start the Oppia server.
```
Can you point to these instructions, or be more specific about what they are?

EricZLou replied:
Yes!

seanlip commented on Dockerfile:
```diff
@@ -0,0 +1,161 @@
+# Build the Docker image from the Oppia root folder (note the "." at the
+# end of the command):
+#
+#   docker build -t {some_name} .
```
Please explain what "some_name" is, here and elsewhere

EricZLou replied:
Can do.

seanlip commented:
Hi @ezl-13 -- could you please only resolve/reply to comments (with "Done" or similar) after you've pushed the code change that addresses them? That will make it easier for reviewers to track things.

If you want to use the comments as a to-do list, you can use the "start a review" button for that. That way, you can create draft replies as you address individual comments, and then publish the whole thing once the code changes are pushed and you're ready for another review. The exception is if you have questions for the reviewer in the interim, in which case please feel free to ask them beforehand as individual comments so that you're not blocked.

Thanks!

seanlip commented:
@ezl-13 -- for the GAE issue, do you think it would be possible to put the Dockerfile in a subdirectory (e.g. windows_installation) and keep it there during the installation process, so as not to introduce conflicts with GAE? I worry that needing to do "rm" commands on various systems is going to interfere  with source control -- the developer will always need to put the file back properly afterwards so that git doesn't detect it as a deletion, and that can become pretty tedious.

EricZLou committed:
Changes to Dockerfile and explanations in karma.conf.ts

EricZLou commented:
Edits made:
- Added links to explain karma.conf.ts flags
- Changed Dockerfile to ubuntu_dockerfile to resolve GAE conflicts
- Edited instructions in dockerfile

PTAL!

EricZLou committed:
Change codeowners

seanlip commented on core/tests/karma.conf.ts:
```diff
@@ -105,8 +106,14 @@ module.exports = function(config) {
+        // Discussion of the necessity of extra flags can be found here:
+        // https://github.com/karma-runner/karma-chrome-launcher/issues/154
+        // https://github.com/karma-runner/karma-chrome-launcher/issues/180
+        flags: isDocker ? ['--no-sandbox',
```
Style nit for this codebase: If something within ( ...) or [ ... ] spans multiple lines, add a line break after the opening paren or brace.

EricZLou replied:
Done!

seanlip commented on ubuntu_dockerfile:
```diff
@@ -0,0 +1,174 @@
+#
+#   docker build -t {some_name} -f ubuntu_dockerfile .
+#      where you should replace {some_name} with whatever you want to call
+#      your Docker image.
```
`your Docker image (say oppia_image)`

and then remove the next line

EricZLou replied:
Got it.

seanlip commented on ubuntu_dockerfile:
```diff
@@ -0,0 +1,174 @@
+# If this is not your first time running this command, add --no-cache flag
+# to rebuild from the beginning:
+#   docker build -t {some_name} -f ubuntu_dockerfile . --no-cache
+#   ex: docker build -t oppia_image -f ubuntu_dockerfile . --no-cache
```
You can delete this line (some_name is clear from above)

EricZLou replied:
Sure.

seanlip commented on ubuntu_dockerfile:
```diff
@@ -0,0 +1,174 @@
+#   ex: docker build -t oppia_image -f ubuntu_dockerfile . --no-cache
+#
+# STEP 4
+# Follow the instructions on how to start the Oppia server below.
```
Delete this line and "step 4", it's superfluous

EricZLou replied:
Done.

seanlip commented on ubuntu_dockerfile:
```diff
@@ -0,0 +1,174 @@
+#   docker run -u 0 -it -p 8181:8181 {some_name}:latest /bin/bash
+#      where you should replace {some_name} with the name of your Docker
+#      image. 
+#   ex: docker run -u 0 -it -p 8181:8181 oppia_image:latest /bin/bash
```
delete this line

EricZLou replied:
:)

seanlip commented on ubuntu_dockerfile:
```diff
@@ -0,0 +1,174 @@
+# This displays a repository name and tag name. Now run:
+#   docker run -u 0 -it -p 8181:8181 {repository}:{tag} /bin/bash
+#      where you should replace {repository} and {tag} with the listed 
+#      name of the repository and tag. 
```
`with the repository and tag names previously listed by "docker ps".`

EricZLou replied:
Done!

seanlip reviewed (changes requested):
Hi @ezl-13 -- final round of comments, otherwise it looks good and I have no concerns. 

Could you please also file follow-up issues (and link to those here) for any remaining work that's needed for a smooth installation using Docker (e.g. for e2e tests and anything else that's remaining to be done as separate projects by either you or @renana)? Thanks!

EricZLou committed:
Final changes

EricZLou commented:
Great! I'll file the issues now and link them here. A Google Doc with instructions to replace the Windows installation page is coming your way @seanlip

seanlip reviewed (approved):
All good from my end, thanks @ezl-13! 

@kevinlee12 @vojtechjelinek  @nithusha21 @DubeySandeep PTAL as codeowners, thanks!

kevinlee12 reviewed (approved):
lgtm!

EricZLou commented:
Issue for getting e2e tests working: https://github.com/oppia/oppia/issues/7345

Issue for Docker Hub integration (automatic update of centralized Docker image): https://github.com/oppia/oppia/issues/7346

Issue for new user-side Dockerfile (pull from Docker Hub and copy user's repo into image): https://github.com/oppia/oppia/issues/7347

EricZLou committed:
Fix typo

EricZLou commented:
Note: I wrote "docker ps" instead of "docker images"
Fixed it in most recent commit. Also, the e2e tests were failing from my last commit, but I'll check what the error is when it finishes running again.

kevinlee12 commented:
@ezl-13, I'm getting an error when I run this locally, here's the dump: https://pastebin.com/zYehpnQ2. FYI I'm running this on Ubuntu now (so this might translate to folks using WSL). I'm going to be running this on my Windows partition later and get you those logs (if it errors).

EricZLou commented:
@kevinlee12 hmm I've never seen that error before. I'll look into it

kevinlee12 commented:
Running the build command on Windows Powershell, I get: 
```
SECURITY WARNING: You are building a Docker image from Windows against a non-Windows Docker host. All files and directories added to build context will have '-rwxr-xr-x' permissions. It is recommended to double check and reset permissions for sensitive files and directories.
```
The command does complete.  I'm currently running the "run" command and will post back when it's done.

kevinlee12 commented:
The good news is that it does start properly on Windows powershell :)

EricZLou commented:
@kevinlee12 Yep! I also get that error. It shouldn't affect anything, and I also mention it in the instructions Google Doc. That's great! Let me know the output of start.sh when it runs.

kevinlee12 commented:
It runs exactly like it does on Linux, with no errors :)

> On Aug 9, 2019, at 11:41 PM, Eric Lou <notifications@github.com> wrote:
> 
> @kevinlee12 Yep! I also get that error. It shouldn't affect anything, and I also mention it in the instructions Google Doc. That's great! Let me know the output of start.sh when it runs.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

EricZLou commented:
Perfect :) and if you get the chance to test frontend tests, let me know how that goes

EricZLou commented:
Note: In my previous commit, about half of the e2e tests failed. 
In my current commit (which had pretty much no change compared to previous commit), 2 e2e tests failed. 

I tried reading through the logs, but I'm still not sure why they're failing. How can I get the tests to rerun? @seanlip

seanlip commented:
Ah, these ones are recent flakiness. I've restarted them. Hopefully they pass!

seanlip commented:
@vojtechjelinek @nithusha21 @DubeySandeep PTAL as codeowners, thanks!

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

kevinlee12 commented:
Running frontend tests:
Run 1 (timeout): https://pastebin.com/kBSrhxE1
Run 2 (timeout): https://pastebin.com/tN49QmhS

For reference, my docker settings are:
![image](https://user-images.githubusercontent.com/5758256/62825264-dddb4c00-bb5d-11e9-9398-62719878c218.png)

EricZLou commented:
@kevinlee12 Yep, this happens sometimes. I also have this in the instructions Google Doc, but please run these 2 commands:
`./node_modules/typescript/bin/tsc --project .
./node_modules/karma/bin/karma start ./core/tests/karma.conf.ts`
Then, if that doesn't work, try the frontend tests script again.

EricZLou commented:
Note: do not merge yet. 
Instructions are changing. Thanks @kevinlee12

EricZLou committed:
Change instructions

EricZLou commented:
Instructions updated in most recent commit.

seanlip commented:
Thanks, @ezl-13! 

@nithusha21 @DubeySandeep PTAL as codeowners!

Also, @kevinlee12 -- just to check, this still LGTM?

kevinlee12 commented:
@ezl-13, running with the node commands, I get: https://pastebin.com/WTepz5KF. Afterwards, I ran the tests using scripts and get https://pastebin.com/TY1rGk59

EricZLou commented:
@kevinlee12 Thanks for testing it out. Manually compiling the ts and starting karma has managed to work for me every time, but I guess the race condition is just in my favor. Either way, I think this PR can be merged because it has reliable functionality for starting the Oppia server. I'll keep working on trying to get run_frontend_tests.sh to work.

seanlip commented:
Hi @ezl-13 -- worth filing an issue for frontend tests?

nithusha21 reviewed (approved):
LGTM as a codeowner. Can you just confirm that all the other environments (Mac OSX and Linux) are unaffected by this change? Other than that LGTM

EricZLou commented:
Issue to make run_frontend_tests.sh script more reliable: https://github.com/oppia/oppia/issues/7355

kevinlee12 commented:
Btw, in light of Nithesh's comment, you should put in the comments that the changes are for Docker only on Windows and not for other platforms.

EricZLou commented:
Yep @nithusha21 the changes are just for Windows users. The only file that I changed that is applicable to Linux / Mac OS users is karma.conf.ts. There is a flag "isDocker" that keeps the old configuration for non Docker users. 

If someone on Linux / Mac OS could verify that the isDocker flag is working as expected, that would be great. All that is required is for you to install Oppia and run the frontend tests script as usual (not via Docker). If it works, then we can be sure that Linux / Mac OS is unaffected.

ezl-13 commented:
@kevinlee12 Can you elaborate on what you mean by "in the comments"?

DubeySandeep reviewed (approved):
LGTM! (As a code owner to the file `.github/CODEOWNERS`)

seanlip commented:
I did the test that @ezl-13 mentioned locally (on Linux) and it works, so I'm merging this.

Thanks @ezl-13!
